lib/src/rules/use_late_for_private_fields_and_variables.dart

```diff
@@ -0,0 +1,179 @@
+    registry.addCompilationUnit(this, visitor);
+  }
+}
+
+class _Visitor extends UnifyingAstVisitor<void> {
+  _Visitor(this.rule, this.context);
+
+  final LintRule rule;
+  final LinterContext context;
+
+  CompilationUnitElement currentUnit;
+
+  static final lateables =
+      <CompilationUnitElement, List<VariableDeclaration>>{};
+  static final nullableAccess = <CompilationUnitElement, Set<Element>>{};
+
+  @override
+  void visitCompilationUnit(CompilationUnit node) {
+    if (node.featureSet.isEnabled(Feature.non_nullable)) {
+      currentUnit = node.declaredElement;
+      lateables.putIfAbsent(currentUnit, () => []);
+      nullableAccess.putIfAbsent(currentUnit, () => {});
+
+      super.visitCompilationUnit(node);
+
+      final unitsInContext =
+          context.allUnits.map((e) => e.unit.declaredElement).toSet();
+      final libraryUnitsInContext = node.declaredElement.library.units
+          .where(unitsInContext.contains)
+          .toSet();
+      final areAllLibraryUnitsVisited =
+          libraryUnitsInContext.every(lateables.containsKey);
+      if (areAllLibraryUnitsVisited) {
+        _checkAccess(libraryUnitsInContext);
+
+        // clean up
+        libraryUnitsInContext.forEach((unit) {
+          lateables.remove(unit);
+          nullableAccess.remove(unit);
+        });
+      }
+    }
+  }
+
+  @override
+  void visitNode(AstNode node) {
+    var element = DartTypeUtilities.getCanonicalElementFromIdentifier(node);
+    if (element != null) {
+      var parent = node.parent;
+      if (parent is Expression) {
+        parent = (parent as Expression).unParenthesized;
+      }
+      if (node is SimpleIdentifier && node.inDeclarationContext()) {
+        // ok
+      } else if (parent is PostfixExpression &&
+          parent.operand == node &&
+          parent.operator.type == TokenType.BANG) {
+        // ok non-null access
+      } else if (parent is AssignmentExpression &&
+          parent.operator.type == TokenType.EQ &&
+          context.typeSystem.isNonNullable(parent.rightHandSide.staticType)) {
+        // ok non-null access
+      } else {
+        nullableAccess[currentUnit].add(element);
+      }
+    }
+    super.visitNode(node);
+  }
+
+  @override
+  void visitFieldDeclaration(FieldDeclaration node) {
+    for (var variable in node.fields.variables) {
+      final parent = node.parent;
+      // see https://github.com/dart-lang/linter/pull/2189#issuecomment-660115569
+      // We could also include public members in private classes but to do that
+      // we'd need to ensure that there are no instances of either the
+      // enclosing class or any subclass of the enclosing class that are ever
+      // accessible outside this library.
+      if (Identifier.isPrivateName(variable.name.name) ||
+          _isPrivateExtension(parent)) {
+        _visit(variable);
+      }
+    }
+  }
+
+  @override
+  void visitTopLevelVariableDeclaration(TopLevelVariableDeclaration node) {
```

Comment:
I believe that this needs to invoke `super.visitTopLevelVariableDeclaration(node)` for a similar reason.